src/pool.js

```diff
@@ -136,35 +143,25 @@ class Pool {
    * @param {Number} [dt] - Time since last update.
    */
   update(dt) {
-    let i = this.size - 1;
     let obj;
-
-    // If the user kills an object outside of the update cycle, the pool won't know of
-    // the change until the next update and this._i won't be decremented. If the user then
-    // gets an object when this._i is the same size as objects.length, this._i will increment
-    // and this statement will evaluate to -1.
-    //
-    // I don't like having to go through the pool to kill an object as it forces you to
-    // know which object came from which pool. Instead, we'll just prevent the index from
-    // going below 0 and accept the fact that this._i may be out of sync for a frame.
-    let index = Math.max(this.objects.length - this._i, 0);
-
-    // only iterate over the objects that are alive
-    while (i >= index) {
+    let doSort = false;
+    let oldSize = this.size;
+    for (let i = this.size; i--; ) {
       obj = this.objects[i];
 
       obj.update(dt);
 
-      // if the object is dead, move it to the front of the pool
       if (!obj.isAlive()) {
-        this.objects = this.objects.splice(i, 1).concat(this.objects);
-        this._i--;
-        index++;
-      }
-      else {
-        i--;
+        doSort = true;
+        this.size--;
       }
     }
+    // sort all dead elements to the end of the pool
+    if (doSort) {
+      let front = this.objects.slice(0, oldSize);
```

Comment:
Interesting. Was there a performance benefit to sorting just the current size and rejoining it together in the end (creating 2 throwaway arrays) over just sorting all the objects?

Reply:
In the quick and simple tests I ran, this was slightly faster, especially in a case where the array is maxed out and the majority of the objects are dead. If a smaller size is a higher priority, this was really splitting hairs on speed.

Reply:
Ya, smaller size is the priority for now.

Reply:
Solved with the inlining of the sort function.